### DISCUSSION TITLE
iOS and Android App - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#2356] [Docs]
# iOS and Android App - Ultralytics YOLOv8 Docs

https://docs.ultralytics.com/app/

**Comment by kodavatimahendra** [comment]
Hi,

Thank you very much for adding the option to comment. 
Can you please add "date" to these posts so that we have an approximate idea of when the page is last updated?

Best regards

**Reply by glenn-jocher** [reply]
@kodavatimahendra this is a great idea! I'll update the Docs with this feature in the next release :)

**Reply by glenn-jocher** [reply]
@kodavatimahendra I've updated all the Docs now to show created and last updated dates! Note all the pages will as recently updated due to the new comments feature addition. 

Let us know if there's anything else you'd like to see :)

**Comment by Artachanidisjohn** [comment]
can i upload my own dataset at the Ultralytics Android App ??

**Reply by glenn-jocher** [reply]
Hi buddy. No, you can not upload a dataset via the app, but you can upload one to HUB and then see it in the app.

More importantly you can train a HUB model on the dataset and then preview it in the app 😃

**Reply by Artachanidisjohn** [reply]
Thank you for your response!!! Have you idea why when i m going to upload the dataset it failed? says Images Not Found (train).  i have zipped, i have the same name with the yaml.

**Reply by glenn-jocher** [reply]
It's probably a format problem. Start by uploading the example coco8.zip and then change it slowly into yours by matching the existing format. If you follow the example coco8 dataset everything should work.

**Comment by markolightdrop** [comment]
Can you share some implementation details regarding this? E.g. which API did you use to deploy yolo on Android and which for iOS?

**Reply by pderrenger** [reply]
@markolightdrop for deploying YOLOv8 on Android and iOS, we typically recommend using export formats that are compatible with mobile deployment frameworks. For Android, you can export the model to TensorFlow Lite (TFLite) format, which is optimized for mobile and edge devices. For iOS, you can export the model to CoreML format, which is designed for high performance on Apple devices.

Here's a brief overview of the process for each platform:

- **Android**: After training your YOLOv8 model, you can export it to ONNX and then convert it to TFLite using TensorFlow's converter tools. The TFLite model can then be integrated into an Android app using the TensorFlow Lite Android Support Library.

- **iOS**: For iOS, you would export your YOLOv8 model to ONNX and then convert it to CoreML using ONNX's CoreML tools. The CoreML model can be integrated into an iOS app using Apple's Core ML framework.

Both of these processes involve a few steps and require some familiarity with the respective mobile development environments. Our documentation provides more detailed guidance on how to perform these exports and integrate the models into mobile apps. Please refer to the Ultralytics Docs for more information on exporting models and deploying them on various platforms. 📱🚀

Keep in mind that deploying deep learning models on mobile devices may require additional optimizations for performance, such as quantization and pruning, due to the limited computational resources available on these devices. 🛠️✨

**Reply by ikalkavouris** [reply]
What do you mean "TensorFlow's converter tools"? TensorFlow does not support officialy ONNX models. Only 3rd party tools can be found here and there on the internet. Can you give a little - just a tiny - bit of more details on the conversion?

PS
Just to quote you:
"Remember, the success of YOLOv8 and its applications is a collective effort, and your feedback helps us improve."
Then share source code and conversion process for the android app.

**Reply by glenn-jocher** [reply]
@ikalkavouris easy! The source code for the Android app is available at https://github.com/ultralytics/yolo-flutter-app :)

**Comment by JoschieB** [comment]
Hi!

First of all, all this stuff is quite amazing, big thanks for all your hard work!

I have 2 questions/remarks

1. I've been training models for my raspberry pi so far, and wanted to also use them on my phone for some quick testing. However on my phone, none of my custom models show any results so far. The standard pre-trained COCO datasets, just work fine and show what it sees. My models show just the camera feed but doesn't see to recognise anything(but it definelty seems like its calculating something). The models all do work on my raspberry, but not on my phone. I am using an Oneplus 8T on Android 13. Any clue what the cause could be?

2. Some of my models don't want to export to android, as in if I click on exporting it will load forever, until after some time it gives the option to retry and that will again load forever. (Of my few models, it seems to happen more on the bigger (m/x) datasets but not consitently)

**Reply by pderrenger** [reply]
@JoschieB hello! Thank you for your kind words and for reaching out with your questions. Let's see if we can resolve the issues you're encountering.

1. If your custom models are working on your Raspberry Pi but not showing results on your phone, there could be several factors at play. It's possible that the model may require additional configuration or optimization for mobile deployment. Ensure that the model is correctly exported for mobile use and that the input data on your phone is preprocessed in the same way as during training. Also, check if there are any version mismatches between the libraries used for training and those on the mobile app.

2. For the issue with exporting models to Android, it sounds like there might be a problem with the export process for larger models. This could be due to memory constraints or other limitations on your device. To troubleshoot, you could try exporting a smaller model to see if the issue persists. Additionally, ensure that you have a stable internet connection during the export process, as interruptions could cause the process to hang.

For both issues, I recommend checking the [Export](https://docs.ultralytics.com/modes/export) and [Predict](https://docs.ultralytics.com/modes/predict) sections of our documentation for detailed guidance on exporting models and making predictions on different platforms. If the problem continues, please open an issue on the GitHub repository with detailed information about the steps you've taken, any error messages, and the specific models you're using, so we can assist you further. 🛠️

Remember, the success of YOLOv8 and its applications is a collective effort, and your feedback helps us improve. Thank you for being part of the YOLO community! 🚀

**Comment by aditcodiant** [comment]
Hi can i count no. of bonding boxes in the app to count the no. of objects

**Reply by pderrenger** [reply]
@aditcodiant yes, you can count the number of bounding boxes to tally the number of detected objects. After running the detection, you can iterate through the results to count the boxes. Each box corresponds to an object that the model has identified. Here's a simplified example of how you might do this in Python:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/your/model.pt')

# Perform detection
results = model.predict('path/to/your/image.jpg')

# Count the number of bounding boxes (objects)
num_objects = len(results.boxes)
print(f'Number of objects detected: {num_objects}')
```

For specific implementations on iOS and Android, you would need to adapt this logic to the respective platform's development environment. 📱🤖

**Comment by lfleon9b** [comment]
Hello, I noticed that the Ultralytics iOS App leverages the Apple Neural Engine and Core ML for real-time YOLO model object detection. Could you please provide possible insights you have into how the performance of YOLOv8 on iOS devices compares to Android devices? Are there specific optimizations or hardware advantages on iOS that contribute to its real-time detection capabilities? Additionally, are there similar optimizations or plans for Android devices to achieve comparable performance? Thanks for any information you can provide

**Reply by pderrenger** [reply]
@lfleon9b hello! 🚀

The performance of YOLOv8 on iOS devices can indeed be quite impressive, thanks in part to the optimization for the Apple Neural Engine and Core ML. These optimizations leverage the hardware acceleration available on iOS devices, which can significantly enhance real-time detection capabilities.

As for Android devices, performance can vary widely due to the diversity of hardware and the absence of a unified acceleration framework like Core ML. However, Android developers can use frameworks like TensorFlow Lite with hardware acceleration options such as NNAPI to optimize for real-time performance.

We are always exploring ways to improve YOLOv8 performance across all platforms, including Android. While specific optimizations may differ due to the underlying hardware and software ecosystems, the goal remains the same: to provide efficient, real-time object detection that can be deployed on a wide range of devices.

For the latest updates and optimizations, keep an eye on our documentation and release notes. Happy coding! 🌟

**Comment by mingdee** [comment]
It's great that I can load my own unlisted model and check it out with this iOS app. However, I want another Ultralytics user to try out the app with my model on their phone. How do I do that?

**Reply by pderrenger** [reply]
@mingdee i'm glad to hear you're interested in sharing your model with another user for use with the Ultralytics app! To share your unlisted model with someone else, you can follow these steps:

1. **Export Your Model**: First, make sure you've exported your YOLOv8 model to the appropriate format that's compatible with the app.

2. **Host Your Model**: Upload your model to a hosting service or server where it can be accessed via a direct download link. Ensure that the hosting service you choose can provide a stable and accessible URL for the model file.

3. **Share the Link**: Share the direct download link to your model with the other user. They will need this link to load the unlisted model within their instance of the Ultralytics app.

4. **Load the Model in the App**: Instruct the other user to open the Ultralytics app on their device, navigate to the section where they can load an unlisted model, and enter the direct download link you provided.

By following these steps, the other user should be able to load and use your model within the Ultralytics app on their iOS or Android device. Remember to ensure that any sharing of models complies with the licensing terms under which the model was created and distributed. 🚀
